internal/cluster/partition_service.go

```diff
@@ -0,0 +1,168 @@
+)
+
+type PartitionServiceCreationBundle struct {
+	SerializationService *iserialization.Service
+	EventDispatcher      *event.DispatchService
+	startCh              chan struct{}
+	startChAtom          int32
+	Logger               ilogger.Logger
+}
+
+func (b PartitionServiceCreationBundle) Check() {
+	if b.SerializationService == nil {
+		panic("SerializationService is nil")
+	}
+	if b.EventDispatcher == nil {
+		panic("EventDispatcher is nil")
+	}
+	if b.Logger == nil {
+		panic("Logger is nil")
+	}
+}
+
+type PartitionService struct {
+	serializationService *iserialization.Service
+	eventDispatcher      *event.DispatchService
+	partitionTable       partitionTable
+	partitionCount       int32
+	logger               ilogger.Logger
+}
+
+func NewPartitionService(bundle PartitionServiceCreationBundle) *PartitionService {
+	bundle.Check()
+	return &PartitionService{
+		serializationService: bundle.SerializationService,
+		eventDispatcher:      bundle.EventDispatcher,
+		partitionTable:       defaultPartitionTable(),
+		logger:               bundle.Logger,
+	}
+}
+
+func (s *PartitionService) Start() {
+	subscriptionID := event.MakeSubscriptionID(s.handlePartitionsUpdated)
+	s.eventDispatcher.Subscribe(EventPartitionsUpdated, subscriptionID, s.handlePartitionsUpdated)
+}
+
+func (s *PartitionService) Stop() {
+	subscriptionID := event.MakeSubscriptionID(s.handlePartitionsUpdated)
+	s.eventDispatcher.Unsubscribe(EventPartitionsUpdated, subscriptionID)
+}
+
+func (s *PartitionService) GetPartitionOwner(partitionId int32) internal.UUID {
+	return s.partitionTable.GetOwnerUUID(partitionId)
+}
+
+func (s *PartitionService) PartitionCount() int32 {
+	return atomic.LoadInt32(&s.partitionCount)
+}
+
+func (s *PartitionService) GetPartitionID(keyData pubserialization.Data) int32 {
+	if count := s.PartitionCount(); count == 0 {
+		// Partition count can not be zero for the sync mode.
+		// On the sync mode, we are waiting for the first connection to be established.
+		// We are initializing the partition count with the value coming from the server with authentication.
+		// This exception is used only for async mode client.
+		// TODO: panic
+		//core.NewHazelcastClientOfflineException()
+		return 0
+	} else {
+		return murmur.HashToIndex(keyData.PartitionHash(), count)
+	}
+}
+
+func (s *PartitionService) GetPartitionIDWithKey(key interface{}) (int32, error) {
+	if data, err := s.serializationService.ToData(key); err != nil {
+		return 0, err
+	} else {
+		return s.GetPartitionID(data), nil
+	}
+}
+
+func (s *PartitionService) handlePartitionsUpdated(event event.Event) {
+	if partitionsUpdatedEvent, ok := event.(*PartitionsUpdated); ok {
+		s.partitionTable.Update(partitionsUpdatedEvent.Partitions, partitionsUpdatedEvent.Version)
+		s.eventDispatcher.Publish(NewPartitionsLoaded())
+		s.logger.Debug(func() string { return "partitions loaded" })
+	}
+}
+
+func (s *PartitionService) checkAndSetPartitionCount(newPartitionCount int32) {
```

Comment:
Even blue/green feature is not implemented, we should check if partition count same. 
If users closes and reopens a cluster ,our clients can connect back to cluster. We don't want our clients to connect to a cluster with different partition count.

Reply:
Updated.

Reply:
@yuce Can you recheck this one ? It seems the method is not still returning an error.

Reply:
@sancar Sorry, this should fix that: https://github.com/yuce/hazelcast-go-client/commit/faafeaa735e6867faeee4a22c4cd3c03384426d9